language/vm/src/serializer.rs

```diff
@@ -32,133 +32,14 @@ where
-    binary: &mut BinaryData,
-    idx: &FunctionInstantiationIndex,
-) -> Result<()> {
-    write_as_uleb128(binary, idx.0, FUNCTION_INST_INDEX_MAX)
-}
-
-fn serialize_struct_def_inst_index(
-    binary: &mut BinaryData,
-    idx: &StructDefInstantiationIndex,
-) -> Result<()> {
-    write_as_uleb128(binary, idx.0, STRUCT_DEF_INST_INDEX_MAX)
-}
-
-fn seiralize_table_offset(binary: &mut BinaryData, offset: u32) -> Result<()> {
-    write_as_uleb128(binary, offset, TABLE_OFFSET_MAX)
-}
-
-fn serialize_table_size(binary: &mut BinaryData, size: u32) -> Result<()> {
-    write_as_uleb128(binary, size, TABLE_SIZE_MAX)
-}
-
-fn serialize_constant_pool_index(binary: &mut BinaryData, idx: &ConstantPoolIndex) -> Result<()> {
-    write_as_uleb128(binary, idx.0, CONSTANT_INDEX_MAX)
-}
-
-fn serialize_bytecode_count(binary: &mut BinaryData, len: usize) -> Result<()> {
-    write_as_uleb128(binary, len as u64, BYTECODE_COUNT_MAX)
-}
-
-fn serialize_identifier_size(binary: &mut BinaryData, len: usize) -> Result<()> {
-    write_as_uleb128(binary, len as u64, IDENTIFIER_SIZE_MAX)
-}
-
-fn serialize_constant_size(binary: &mut BinaryData, len: usize) -> Result<()> {
-    write_as_uleb128(binary, len as u64, CONSTANT_SIZE_MAX)
-}
-
-fn serialize_field_count(binary: &mut BinaryData, len: usize) -> Result<()> {
-    write_as_uleb128(binary, len as u64, FIELD_COUNT_MAX)
-}
-
-fn serialize_field_offset(binary: &mut BinaryData, offset: u16) -> Result<()> {
-    write_as_uleb128(binary, offset, FIELD_OFFSET_MAX)
-}
-
-fn serialize_acquires_count(binary: &mut BinaryData, len: usize) -> Result<()> {
-    write_as_uleb128(binary, len as u64, ACQUIRES_COUNT_MAX)
-}
-
-fn serialize_signature_size(binary: &mut BinaryData, len: usize) -> Result<()> {
-    write_as_uleb128(binary, len as u64, SIGNATURE_SIZE_MAX)
-}
-
-fn serialize_type_parameter_index(binary: &mut BinaryData, idx: u16) -> Result<()> {
-    write_as_uleb128(binary, idx, TYPE_PARAMETER_INDEX_MAX)
-}
-
-fn serialize_type_parameter_count(binary: &mut BinaryData, len: usize) -> Result<()> {
-    write_as_uleb128(binary, len as u64, TYPE_PARAMETER_COUNT_MAX)
-}
-
-fn serialize_bytecode_offset(binary: &mut BinaryData, offset: u16) -> Result<()> {
-    write_as_uleb128(binary, offset, BYTECODE_INDEX_MAX)
-}
-
-fn serialize_table_count(binary: &mut BinaryData, len: u8) -> Result<()> {
-    write_as_uleb128(binary, len, TABLE_COUNT_MAX)
-}
-
-fn serialize_local_index(binary: &mut BinaryData, idx: u8) -> Result<()> {
-    write_as_uleb128(binary, idx, LOCAL_INDEX_MAX)
-}
-
 impl CompiledScriptMut {
     /// Serializes this into a binary format.
     ///
     /// This is intended mainly for test code. Production code will typically use
     /// [`CompiledScript::serialize`].
     pub fn serialize(&self, binary: &mut Vec<u8>) -> Result<()> {
         let mut binary_data = BinaryData::from(binary.clone());
-        let mut ser = ScriptSerializer::new(1);
+        let mut ser = ScriptSerializer::new(VERSION_1);
```

Comment:
Why don't we just serialize it to the version we deserialized it from?

Reply:
You mean, the latest version?

Reply:
No, I mean whatever we deserialized this from. We are deserializing it from Version X, why are we not re-serializing it back as Version X? Why is the version hardcoded?

Reply:
I guess partially the reason is, we did not record the version number in `CompiledModule` :(

Reply:
Given `CompiledScriptMut`, there doesn't necessarily exist a version X blob it was serialized to, right? I mean the script could have been created entirely in memory, for example, by the compiler.

Reply:
Right, so we have to choose a version to serialize to.